12002 inst ref message parameters

amullarney opened the pull request:
Provide for inter-xtUML-component messaging to include parameters of type inst_ref and inst_ref_set. Such messages are serialized on the sending side and de-serialized by the receiver.

amullarney committed:
Compiles inst_ref test

Only casts incoming inst_ref params.. why? - should be both directions.
Needs inst_ref_set serialize/deserialize

amullarney committed:
Add inst_ref cast to outgoing msg parms.

Moving on to inst_ref_set.

amullarney committed:
Set params serialization

I/F files for ClassSet need work.
But, passing instances in a set is working.

amullarney committed:
Inst and Set params

Test case successfully passes instance and instance set.

amullarney committed:
Saving minor changes

Template naming, consistency for class/set templates.
All non functional.

amullarney committed:
Add exception

Sanity check with CarPark revealed lack of XtumlException needed by new
'serialize'method.
Otherwise, non-functional comments.

johnrwolfe commented on src/tool-core/models/tool-core/architecture/classes/Inst Set/Inst Set.xtuml:
```diff
@@ -55,12 +65,21 @@ end for;
 T::clear();
 
+// Determine whether the corresponding class is defined outside a component
+// If so, allow this interface to extend the corresponding class set independent interface.
```
Should this be "...corresponding component-independent class set interface."?

johnrwolfe commented on src/tool-core/models/tool-core/architecture/interfaces/Message/Message.xtuml:
```diff
@@ -246,6 +255,17 @@ INSERT INTO EP_PKG_PROXY
 	'interfaces',
-	'',
+	'Interfaces support passing instances and sets of instances as parameters on Port Messages.
+This is only practical because components can share component independent definitions of
```
While this is true for two xtUML components passing inst_ref and inst_ref_set as message parameters to one another, it is not required for an xtUML component interacting with external code (e.g., a web client).

johnrwolfe commented on src/tool-core/models/tool-core/architecture/interfaces/Message/Message.xtuml:
```diff
@@ -246,6 +255,17 @@ INSERT INTO EP_PKG_PROXY
+This is only practical because components can share component independent definitions of
+classes - and, therefore, sets of class instances - through the Package Reference mechanism.
+
+While shared attribute, ssociation and state model definitions of such classes are immutable, 
```
Typo:  "ssociation"
Consider the following rewording of the second paragraph:
While shared attribute, association and state model definitions of such classes are immutable, 
within components shared classes may participate in component-specific associations.

johnrwolfe commented on src/tool-core/models/tool-core/architecture/interfaces/Message/Message.xtuml:
```diff
@@ -246,6 +255,17 @@ INSERT INTO EP_PKG_PROXY
+While shared attribute, ssociation and state model definitions of such classes are immutable, 
+within components the classes can be extended with component-specific associations.
+
+Inter-component messages must use component-independent definitions of classes used as 
```
Consider the following rewording of the third paragraph:
Since each component-specific class interface includes methods for all associations, including component-specific ones, in which the class participates, 
each method representing an inter-component message must use component-independent interfaces to declare inst_ref and inst_ref_set parameters.  Therefore, methods realizing port-message activities must also use component-independent interfaces when declaring such parameters which are then down-cast to the receiving-component-specific class or class set required by the body of the port-message activity.

johnrwolfe commented on src/tool-core/models/tool-core/architecture/interfaces/Message/Message.xtuml:
```diff
@@ -14,12 +14,16 @@ INSERT INTO O_TFR
-	'// Generate the interface methods to create Port Messages.
-// An Interface mark may specify message parameters passed as JSON strings, rather than indexed array.
-// Parameter-handling for both mechanisms are built; one is then chosen based on that marked value.
+	'// Render the Interface file methods which create Port Messages.
```
Wording is confusing.  Maybe it should be something like this:
Render methods which create instances of Port Messages.

johnrwolfe commented on src/tool-core/models/tool-core/architecture/interfaces/Message/Message.xtuml:
```diff
@@ -14,12 +14,16 @@ INSERT INTO O_TFR
-// An Interface mark may specify message parameters passed as JSON strings, rather than indexed array.
-// Parameter-handling for both mechanisms are built; one is then chosen based on that marked value.
+	'// Render the Interface file methods which create Port Messages.
+// Note that any parameters of type inst_ref or inst_ref_set must be fully qualified as component-independent.
```
In this comment, please explain why the fully-qualified name of the component-independent type is used instead of requesting a type reference for it with a goal of producing an import statement for the component-independent type.

johnrwolfe commented on src/tool-core/models/tool-core/architecture/interfaces/Message/Message.xtuml:
```diff
@@ -14,12 +14,16 @@ INSERT INTO O_TFR
+// Note that any parameters of type inst_ref or inst_ref_set must be fully qualified as component-independent.
+// An Interface ''mark'' may specify message parameters be passed as key-value pair strings, rather than an indexed array.
+// Additionally, discovery of an inst_ref/set parameter type triggers message serialization.
+// Parameter-handling for both mechanisms (array/string) are prepared; the appropriate one is then implemented.
```
Parameter-handling methods for both...

johnrwolfe reviewed (changes requested):
A few questions and comments are embedded below.

amullarney committed:
Limit serialization

Only serialize classes/sets used as messgae params.
Reworked descriptions.

amullarney committed:
Complete

Amplified attribute description.

amullarney committed:
Derived attribute serialize

Use get accessors in serialize

amullarney committed:
Class/set as port msg param

Test for Port message param at class/set render.

amullarney committed:
Attribute deserialize

Use type-specific JSON get methods.

amullarney committed:
Remove unreferenced type

Remove unreferenced type: component-independent class.
This is never directly called for by type_reference.
Used only in 2 contexts: Interface & Port Message.
In those, the reference is explicitly constructed.

Minor changes to selector render() for clarity.
Tested: CarPark; Payroll, 12002 AdHoc.
Also rebuilds itself successfully.
